File: find_min_max.cpp
//Time complexity - 3 comparisions are being done here within O(n)
//Space complexity - O(1)

#include <bits/stdc++.h>
using namespace std;

int main() {
	// your code goes here
    int A[] = {};
    if(A.empty())
    {
        cout<<"empty array"<<endl;
        return 0;
    }
    int n=A.size();
    int min=INT_MAX,max=INT_MIN;
    int imin,imax;
    for(int i=0;i<n;i+=2)
    {
        if(A[i]<A[i+1])
        {
            imin=A[i];
            imax=A[i+1];
        }
        else
        {
            imax=A[i];
            imin=A[i+1];
        }
        if(imin<min)
            min=imin;
        if(imax>max)
            max=imax;
    }
    cout<<"Minimum element is "<<min; 
    cout<<"nMaximum element is "<<max; 
	return 0;
}

File: find_min_max.py
# Time Complexity : O(n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach in three sentences only
"""
In this code we try to get the minimum and the maximum value of an array by comparing the elements in pairs.
"""

class Solution():
    def get_min_max(self, arr, low, high): # with minimum comparisions through tournaments.
        if len(arr[low:high+1]) == 1:
            return [arr[low], arr[low]]
        
        elif len(arr[low:high+1]) == 2:
            return [arr[low], arr[high]] if arr[low] < arr[high] else [arr[high], arr[low]]
        
        else:
            mid = (low+high)//2
            min1, max1 = self.get_min_max(arr, low, mid)
            min2, max2 = self.get_min_max(arr, mid+1, high)

            return [min(min1, min2), max(max1, max2)]

arr = [1000, 11, 445, 1, 330, 3000]

sol = Solution()
print(sol.get_min_max(arr, 0, len(arr)-1))
